Fixes #14399 - fix error on inc update with composites

jlsherrill opened the pull request.

jlsherrill commented:
Also adds support for doing an incremental update without specifying an environment

beav commented:
@jlsherrill a newbie question, but I am having some trouble getting the incremental update button to appear in the web ui to test this PR. How do you usually kick off the incremental update?

bbuckingham commented:
@beav, The incremental update is initiated through the Content -> Errata page.

The following is an example scenario:
- go to Content -> Errata
- select an applicable errata
- select Content Hosts tab
- select a host
- select Apply to Hosts
- select 'Apply to Content Host Immediately after publishing'
- click Confirm

I had tested out a couple of the other bug fixes for incremental updates, so I can try this one out as well.

beav commented:
@bbuckingham thanks for the info, and for covering this PR :koala:

bbuckingham commented:
I have tested this PR using the scenario described above; however, I am not seeing the expected behavior based on interpretation of original bug description.  

One note regarding the scenario tested, the content host was registered to a component view (e.g. zoo 1.0) that needed an incremental update.  That component view (zoo 1.0) was also part of a published composite view (e.g. composite 1.0).  

In discussing the original bug with QE, the assumption was that performing an incremental update on the component (zoo 1.0), would also propagate the incremental update to the composite; however, I do not see this occuring.  The following is an example of how this could be done with the cli:
```
hammer> content-view version incremental-update --lifecycle-environment-ids 2 --content-view-version-id 12 --propagate-all-composites true --puppet-module-ids 1

```

jlsherrill commented:
@bbuckingham i'm seeing it work for me, did you use this PR:  https://github.com/Katello/hammer-cli-katello/pull/376 as well?  Apologies for not including it in my original description.  

With those two prs i'm seeing:

```
[vagrant@robot hammer-cli-katello]$ hammer -u admin -p changeme  content-view version incremental-update --lifecycle-environment-ids 1 --content-view-version-id 57  --propagate-all-composites true  --puppet-module-ids=1
[.................................................................................................................................] [100%]
Content View: puppet-component version 1.1
Added Content:
  Puppet Modules:
        puppetlabs-apache-1.3.0

Content View: puppet-composite version 1.1
Added Content:
  Puppet Modules:
        puppetlabs-apache-1.3.0
```

jlsherrill commented:
@bbuckingham actually that pr shouldn't be required (its only required if you don't want to specify an environment id which has nothing to do with composites).  I can't seem to reproduce the issue, can you?

jlsherrill commented on app/lib/actions/katello/content_view_version/incremental_update.rb:
```diff
@@ -108,7 +108,7 @@ def finalize
           old_components = old_version.components.select do |component|
-            !old_version.components.map(&:content_view_id).include?(component.content_view_id)
+            !new_components.map(&:content_view_id).include?(component.content_view_id)
           end
```
@bbuckingham this is the change from what we just discovered.  If you want me to walk you through what this is doing and explain why it was wrong, i'd be happy to!

bbuckingham replied:
@jlsherrill, With this change, I am now able to perform incremental update on a content host registered to a composite view (with multiple components).

bbuckingham commented:
ACK
